test/Sema/implementation-only-import-in-decls.swift

```diff
@@ -0,0 +1,147 @@
+public var testBadTypeInferred = Optional<BadStruct>.none // expected-error {{cannot use 'BadStruct' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+public var testBadTypePartiallyInferred: Optional = Optional<BadStruct>.none // expected-error {{cannot use 'BadStruct' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+public var (testBadTypeTuple1, testBadTypeTuple2): (BadStruct?, BadClass?) = (nil, nil)
+// expected-error@-1 {{cannot use 'BadStruct' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+// expected-error@-2 {{cannot use 'BadClass' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+public var (testBadTypeTuplePartlyInferred1, testBadTypeTuplePartlyInferred2): (Optional, Optional) = (Optional<Int>.none, Optional<BadStruct>.none) // expected-error {{cannot use 'BadStruct' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+
+extension BadStruct { // expected-error {{cannot use 'BadStruct' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+  public func testExtensionOfBadType() {} // FIXME: Should complain here instead of at the extension decl.
+}
+extension BadStruct {
+  func testExtensionOfOkayType() {}
+}
+
+extension Array where Element == BadStruct { // expected-error {{cannot use 'BadStruct' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+  public func testExtensionWithBadRequirement() {} // FIXME: Should complain here instead of at the extension decl.
+}
+
+extension Array where Element == BadStruct {
+  func testExtensionWithOkayRequirement() {} // okay
+}
+
+extension Int: BadProto {} // expected-error {{cannot use 'BadProto' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+struct TestExtensionConformanceOkay {}
+extension TestExtensionConformanceOkay: BadProto {} // okay
+
+public protocol TestConstrainedExtensionProto {}
+extension Array: TestConstrainedExtensionProto where Element == BadStruct { // expected-error {{cannot use 'BadStruct' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+}
+
+
+infix operator !!!!!!: BadPrecedence // expected-error {{cannot use 'BadPrecedence' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+
+precedencegroup TestLowerThan {
+  lowerThan: BadPrecedence // expected-error {{cannot use 'BadPrecedence' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+}
+precedencegroup TestHigherThan {
+  higherThan: BadPrecedence // expected-error {{cannot use 'BadPrecedence' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+}
+
+public struct PublicStructStoredProperties {
+  public var publiclyBad: BadStruct? // expected-error {{cannot use 'BadStruct' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
+  internal var internallyBad: BadStruct? // expected-error {{cannot use 'BadStruct' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
```

Comment:
I wish the diagnostic was more clear about why it can't be used there ("cannot use 'BadStruct' in a public declaration", "cannot use 'BadStruct' in a stored property", etc.). But if you want to wait until we have a proper user-facing feature, I suppose that's acceptable.

Reply:
Yeah, I think we can do something more like the access control diagnostics later.